### ISSUE TITLE
SVM Formulation

### ISSUE THREAD

**Issue by AidanGG** [AidanGG/TensorFlow-to-TMVA-Adapter#1]
From what I know, TensorFlow doesn't handle constrained optimisation easily. Therefore, I would want an unconstrained support vector machine (SVM) cost function.

Indeed, I could minimise some unconstrained cost function in the primal formulation:
![equation](https://latex.codecogs.com/gif.latex?W%20%3D%20%5Cfrac%7B1%7D%7B2%7D%20%5Cleft%20%5C%7C%20%5Cvec%7Bw%7D%20%5Cright%20%5C%7C%5E%7B2%7D%20&plus;%20C%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Cxi%20_%7Bi%7D),
which lacks kernelising and support for non-linear SVMs.

Alternatively, I could maximise the dual formulation:
![equation](https://latex.codecogs.com/gif.latex?L%3D%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Calpha_%7Bi%7D-%5Cfrac%7B1%7D%7B2%7D%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Csum_%7Bj%3D1%7D%5E%7Bn%7D%5Calpha_%7Bi%7D%5Calpha_%7Bj%7Dy_%7Bi%7Dy_%7Bj%7DK%28%5Cvec%7Bx_%7Bi%7D%7D%2C%5Cvec%7Bx_%7Bj%7D%7D%29),
subject to the constraints
![equation](https://latex.codecogs.com/gif.latex?%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Calpha_%7Bi%7Dy_%7Bi%7D%3D0)
and
![equation](https://latex.codecogs.com/gif.latex?0%5Cleq%20%5Calpha%20%5Cleq%20C)
which does support kernelising, but has constraints.

I am looking for a way that allows kernelising, without optimisation constraints.

**Comment by AidanGG** [comment]
According to this paper, Section 2.3:
http://www.kyb.mpg.de/fileadmin/user_upload/files/publications/attachments/primal_%5b0%5d.pdf,
equation 2.13 suggests a way to do this. However, the arithmetic they use neglects the offset term. Though it shouldn't be to difficult to derive an equivalent expression that incorporates the offset term, I would have to do more research to ensure that it is rigorous.

**Comment by AidanGG** [comment]
Another idea I had was to combine one of the constraints in, so that I would minimise
![equation](https://latex.codecogs.com/gif.latex?M%5Cleft%20%28%20%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Calpha%20_%7Bi%7Dy_%7Bi%7D%20%5Cright%20%29-L)
for some large constant ![equation](https://latex.codecogs.com/gif.latex?M), to force the constraint to be preferentially satisfied. I could then incorporate the box constraint by subclassing an optimiser in TensorFlow.

However, this approach seems inelegant, and I am hesitant to implement it.

**Comment by AidanGG** [comment]
Perhaps implementing a double optimiser, one for the dual formulation, and another for its constraints would work. I would first apply gradient ascent on ![equation](https://latex.codecogs.com/gif.latex?L), and then gradient descent on
![equation](https://latex.codecogs.com/gif.latex?%5Cleft%20%7C%5Csum_%7Bi%3D1%7D%5E%7Bn%7D%5Calpha%20_%7Bi%7Dy_%7Bi%7D%20%5Cright%20%7C)
where I also implement the box constraints.

**Comment by AidanGG** [comment]
Looking again at the paper's solution, I am inclined to think that the expression that I'm looking for is simply just
![equation](https://latex.codecogs.com/gif.latex?%5Cboldsymbol%7B%5COmega%20%7D%28%5Cboldsymbol%7B%5Cbeta%7D%2C%20b%29%3A%3D%20%5Clambda%5Cboldsymbol%7B%5Cbeta%5E%7B%5Ctop%7D%7DK%5Cboldsymbol%7B%5Cbeta%7D&plus;%5Csum_%7Bi%3D1%7D%5E%7Bn%7DL%28y_%7Bi%7D%2CK_%7Bi%7D%5E%7B%5Ctop%7D%5Cboldsymbol%7B%5Cbeta%7D&plus;b%29).
I will implement this solution and come back to it if it does not prove suitable.

**Comment by AidanGG** [comment]
I have implemented the solution from the last comment in the develop branch.

**Comment by AidanGG** [comment]
This is the correct solution, after confirming with the author of the paper.
